crates/astria-sequencer/src/consensus.rs

```diff
@@ -0,0 +1,144 @@
+            let _ = rsp_sender.send(Ok(match req {
+                ConsensusRequest::InitChain(init_chain) => ConsensusResponse::InitChain(
+                    self.init_chain(init_chain)
+                        .await
+                        .expect("init_chain must succeed"),
+                ),
+                ConsensusRequest::BeginBlock(begin_block) => ConsensusResponse::BeginBlock(
+                    self.begin_block(begin_block)
+                        .await
+                        .expect("begin_block must succeed"),
+                ),
+                ConsensusRequest::DeliverTx(deliver_tx) => {
+                    ConsensusResponse::DeliverTx(self.deliver_tx(deliver_tx).await?)
+                }
+                ConsensusRequest::EndBlock(end_block) => ConsensusResponse::EndBlock(
+                    self.end_block(end_block)
+                        .await
+                        .expect("end_block must succeed"),
+                ),
+                ConsensusRequest::Commit => {
+                    ConsensusResponse::Commit(self.commit().await.expect("commit must succeed"))
+                }
+            }));
+        }
+        Ok(())
+    }
+
+    async fn init_chain(
+        &mut self,
+        init_chain: request::InitChain,
+    ) -> Result<response::InitChain, BoxError> {
+        // the storage version is set to u64::MAX by default when first created
+        if self.storage.latest_version() != u64::MAX {
+            return Err(anyhow!("database already initialized").into());
+        }
+
+        let genesis_state: GenesisState = serde_json::from_slice(&init_chain.app_state_bytes)
+            .expect("can parse app_state in genesis file");
+
+        self.app.init_chain(&genesis_state).await?;
+
+        // TODO: return the genesis app hash
+        Ok(Default::default())
+    }
+
+    async fn begin_block(
+        &mut self,
+        begin_block: request::BeginBlock,
+    ) -> Result<response::BeginBlock, BoxError> {
+        if self.storage.latest_version() == u64::MAX {
+            // TODO: why isn't tendermint calling init_chain before the first block?
+            self.app
+                .init_chain(&GenesisState::default())
+                .await
+                .expect("init_chain must succeed");
+        }
+
+        let events = self.app.begin_block(&begin_block).await;
+        Ok(response::BeginBlock {
+            events,
+        })
+    }
+
+    async fn deliver_tx(
+        &mut self,
+        deliver_tx: request::DeliverTx,
+    ) -> Result<response::DeliverTx, BoxError> {
+        self.app
+            .deliver_tx(&deliver_tx.tx)
+            .await
+            .unwrap_or_else(|e| {
+                // we don't want to panic on failing to deliver_tx as that would crash the entire
+                // node
+                tracing::error!(error = ?e, "deliver_tx failed");
+                vec![]
+            });
+        Ok(Default::default())
```

Comment:
Should this not return a `BoxError` if `self.app.deliver_tx` fails?

Reply:
I think I changed this because a failing `deliver_tx` (ie meaning the tx execution failed) would cause the entire blockchain to halt, as it wouldn't be able to make a block. I'll think about how to fix that, it should still do something different on error

Reply:
I see - note that this makes it an infallible operation (as are many of the other async methods in this impl block).

I think I am fine with these functions being handlers that only return the final result `T` (as opposed to a `Result<T, E>` where, if `Err(e)` is encountered, the event loop calling `deliver_tx` makes a decision on what to return).

The only case where this could be different is if an invariant is violated and the event loop should exit.